### PR TITLE
Record max concurrency in MetricItem and metric logs

### LINKED ISSUE #349: Record max concurrency in MetricItem and metric logs
## Issue Description

Type: *feature request*

### Describe what feature you want

Record max concurrency in `MetricItem` and metric logs. The `Concurrency` attribute in `MetricItem` was reserved for this.

### PR DESCRIPTION
**Pull request by liqiangz**
### Describe what this PR does / why we need it
Record max concurrency in MetricItem and metric logs

### Does this pull request fix one issue?

Fixes #349
### Special notes for reviews

### CODE REVIEW

File: core/base/constant.go [hunk 1]
```diff
@@ -29,4 +29,6 @@ const (
 	DefaultIntervalMsTotal uint32 = 10000
 
 	DefaultStatisticMaxRt = int64(60000)
+
+	DefaultStatisticMinConcurrency = int32(0)
```
Comment (louyuting) [inline core/base/constant.go:33]: Maybe :`DefaultStatisticMaxConcurrency `

Comment (liqiangz) [reply core/base/constant.go:33]: I mainly refer to the default value naming of the minRt() method here, and here is mainly to calculate the maximum value, then the default value should be the minimum value. So the name is DefaultStatisticMinConcurrency.

Comment (sczyh30) [reply core/base/constant.go:33]: Maybe the constant `DefaultStatisticMinConcurrency` is not necessary, as the initial concurrency should be 0 apparently.

Comment (liqiangz) [reply core/base/constant.go]: done

File: core/stat/base/metric_bucket.go [hunk 2]
```diff
@@ -47,6 +49,10 @@ func (mb *MetricBucket) Add(event base.MetricEvent, count int64) {
 		mb.AddRt(count)
 		return
 	}
+	if event == base.MetricEventConcurrency {
```
[the remark this answers is not in the document]
Comment (liqiangz) [reply core/stat/base/metric_bucket.go]: done

File: core/stat/base_node.go [hunk 3]
```diff
@@ -80,12 +84,17 @@ func (n *BaseStatNode) MinRT() float64 {
...
+func (n *BaseStatNode) MaxConcurrency() int32 {
+	return n.metric.MaxConcurrency()
+}
+
 func (n *BaseStatNode) CurrentConcurrency() int32 {
 	return atomic.LoadInt32(&(n.concurrency))
 }
 
 func (n *BaseStatNode) IncreaseConcurrency() {
 	atomic.AddInt32(&(n.concurrency), 1)
```
Comment (louyuting) [inline core/stat/base_node.go:96]: Use return value (`atomic.AddInt32(&(n.concurrency), 1)`) to  update concurrency

Comment (liqiangz) [reply core/stat/base_node.go]: done

**Review by luckyxiaoqiang — APPROVED** [review]
LGTM

**Review by sanxun0325 — APPROVED** [review]
LGTM

File: core/stat/base/metric_bucket.go [hunk 4]
```diff
@@ -26,13 +26,15 @@ import (
...
+	counter        [base.MetricEventTotal]int64
+	minRt          int64
+	maxConcurrency int32
 }
 
 func NewMetricBucket() *MetricBucket {
 	mb := &MetricBucket{
-		minRt: base.DefaultStatisticMaxRt,
+		minRt:          base.DefaultStatisticMaxRt,
+		maxConcurrency: int32(0),
```
[the remark this answers is not in the document]
Comment (liqiangz) [reply core/stat/base/metric_bucket.go]: done

File: core/stat/base/metric_bucket.go [hunk 5]
```diff
@@ -26,13 +26,15 @@ import (
 // Note that all operations of the MetricBucket are required to be thread-safe.
 type MetricBucket struct {
 	// Value of statistic
-	counter [base.MetricEventTotal]int64
-	minRt   int64
+	counter        [base.MetricEventTotal]int64
+	minRt          int64
+	maxConcurrency int32
```
[the remark this answers is not in the document]
Comment (liqiangz) [reply core/stat/base/metric_bucket.go:31]: Non-negative numbers are dangerous in subtraction

**Review by luckyxiaoqiang — APPROVED** [review]

**Pull request merged**